Source/WebCore/rendering/RenderElement.cpp

```diff
@@ -2487,7 +2487,7 @@ bool RenderElement::hasEligibleContainmentForSizeQuery() const
 void RenderElement::clearNeedsLayoutForSkippedContent()
 {
     for (CheckedRef descendant : descendantsOfType<RenderObject>(*this))
-        descendant->clearNeedsLayout(EverHadSkippedContentLayout::No);
+        descendant->clearNeedsLayoutForSkippedContent();
```

Comment:
This is not good. It will be highly inefficient if one of the descendants is itself a `RenderElement`. We will loop over the same `RenderObject` one time for each `RenderElement` enclosing it. This should be left as-is.

Reply:
Besides Darin's comment, is the ASSERT the same as described in the bug? I landed some PRs that were expected to avoid this dirty renderer problem. If so, which RenderElement is not laid out? Extra points if you can show a render tree dump (showRenderTreeForThis).

Reply:
> Besides Darin's comment, is the ASSERT the same as described in the bug?

Yes, it is:
```
ASSERTION FAILED: !renderer.needsLayout()
/Users/vitaly/Developer/WebKit/Source/WebCore/rendering/RenderBlock.cpp(3174) : void WebCore::RenderBlock::checkPositionedObjectsNeedLayout()
com.apple.WebKit.WebContent.Development terminated (pid 39643) for reason: crash
```

Reply:
Do you know which renderer has dirty layout? 
I think my fixes were for out-of-flow positioned content, if the test does not contain out-of-flow content, and it does not look like it has, then that is an obvious reason my previous fixes would not work here.

Reply:
> Extra points if you can show a render tree dump (showRenderTreeForThis).

```
(B)lock/(I)nline/I(N)line-block, (A)bsolute/Fi(X)ed/(R)elative/Stic(K)y, (F)loating, (O)verflow clip, Anon(Y)mous, (G)enerated, has(L)ayer, hasLayer(S)crollableArea, (C)omposited, Content-visibility:(H)idden/(A)uto, (S)kipped content, (+)Dirty style, (+)Dirty layout
B---YGLSC-- --  RenderView at (0,0) size 800x600 renderer (0x1150002c0) layout box (0x0)
B-----LSC-- --    HTML RenderBlock at (0,0) size 800x222 renderer (0x115001700) layout box (0x0) node (0x115001050)
B---------- --      BODY RenderBody at (8,8) size 784x206 renderer (0x115001a20) layout box (0x11500ea20) node (0x115001250)
---------- --        line at (0.00,0.00) size (784.00x206.00) baseline (202.00) enclosing top (0.00) bottom (202.00)
---------- --          Root inline box at (0.00,188.00) size (202.00x18.00)
---------- --          Run(s):
---------- --            Atomic box at (0.00,0.00) size 202.00x202.00 renderer->(0x11500e810)
I-----L-CH- --        VIDEO RenderVideo at (0,0) size 202x202 renderer (0x11500e810) layout box (0x11500ead0) node (0x115005550)
BR----L---S -+          DIV RenderBlock at (1,1) size 200x0 renderer (0x115003630) layout box (0x0) node (0x115006350) layout->[normal child][positioned child]
BA----L---S -+            DIV RenderBlock at (0,0) size 200x200 renderer (0x1150038d0) layout box (0x0) node (0x115006530) layout->[self][normal child][positioned child]
BA----L---S --              BUTTON RenderFlexibleBox at (100,100) size 60x60 renderer (0x115011af0) layout box (0x0) node (0x115006810)
BA----L---S -+*               DIV RenderBlock at (0,0) size 60x60 renderer (0x115011ea0) layout box (0x0) node (0x11500cd20) layout->[self][normal child]
BA----L---S -+                  DIV RenderBlock at (0,0) size 60x60 renderer (0x115012140) layout box (0x0) node (0x11500cda0) layout->[self]
BA----L---S -+                  DIV RenderBlock at (0,0) size 60x60 renderer (0x1150123e0) layout box (0x0) node (0x11500ce20) layout->[self]
BA----L---S -+                PICTURE RenderBlock at (3,0) size 60x60 renderer (0x115012680) layout box (0x0) node (0x1150069d0) layout->[self]
I---------- --        #text RenderText renderer (0x115011550) layout box (0x115011620) node (0x115011310) length->(3) "\n\n\n"
I---------- --        #text RenderText renderer (0x115011800) layout box (0x1150116c0) node (0x1150113f0) length->(2) "\n\n"
```

Reply:
> This is not good. It will be highly inefficient if one of the descendants is itself a `RenderElement`. We will loop over the same `RenderObject` one time for each `RenderElement` enclosing it. This should be left as-is.

You're right. I've missed the fact that we're traversing all descendants. We should use `childrenOfType` here.

Reply:
Thanks! I lack time to investigate, maybe end of the week I have some time. If you have some cycles before that, I would point out that RenderVideo is inline rather than block, and also not a RenderBox, and I wonder if the content-visibility logic does enough clearing of the layout bit in this case.